File: app/src/main/java/com/foxtrader/app/domain/usecase/ai/agents/NewsAgent.kt
package com.foxtrader.app.domain.usecase.ai.agents

import com.foxtrader.app.domain.model.AgentContext
import com.foxtrader.app.domain.model.AgentInsight
import com.foxtrader.app.domain.model.AgentName
import com.foxtrader.app.domain.model.AgentOutput
import com.foxtrader.app.domain.model.AgentStatus
import com.foxtrader.app.domain.model.Bias
import com.foxtrader.app.domain.usecase.ai.TradingAgent
import javax.inject.Inject

/**
 * NEWS agent — assesses whether high-impact news is imminent and can BLOCK
 * if within a blackout window. Runs in phase 1 (foundational).
 *
 * Inputs via [AgentContext]: `minutesToHighImpactNews` (null = unknown / no
 * calendar), `inNewsBlackout` (true when manually set by user or by the news
 * engine detecting a red-flag event within the blackout window).
 *
 * Behavior:
 * - Blackout active OR < 15 min to a red event → emits a BLOCK insight.
 * - 15..60 min out → caution (NEUTRAL, no block).
 * - No upcoming news or > 60 min → NEUTRAL, no insights.
 */
class NewsAgent @Inject constructor() : TradingAgent {

    override val name = AgentName.NEWS
    override val description = "High-impact news proximity — can BLOCK during blackout windows."
    override val version = "1.0.0"

    override fun analyze(context: AgentContext): AgentOutput {
        val start = System.nanoTime()
        val now = System.currentTimeMillis()

        // Explicit blackout flag.
        if (context.inNewsBlackout) {
            return blockOutput("News blackout active — stand aside.", start, now)
        }

        val minutes = context.minutesToHighImpactNews
        if (minutes == null) {
            // No calendar data → cannot assess; stay neutral.
            return neutralOutput(name, "No news calendar data available.", start)
        }

        return when {
            minutes < BLACKOUT_MINUTES -> blockOutput(
                "High-impact news in $minutes min — inside blackout window ($BLACKOUT_MINUTES min).",
                start, now
            )
            minutes < CAUTION_MINUTES -> AgentOutput(
                agentName = name,
                status = AgentStatus.COMPLETE,
                bias = Bias.NEUTRAL,
                confidence = 40.0,
                insights = listOf(
                    AgentInsight(
                        id = "$name-CAUTION",
                        agentName = name,
                        type = "NEWS_CAUTION",
                        direction = null,
                        confidence = 40.0,
                        timestamp = now,
                        detail = "High-impact news in $minutes min — caution, do not overcommit.",
                        weight = 0.5,
                        tags = listOf("NEWS_CAUTION"),
                    )
                ),
                narrative = "News in $minutes min — caution.",
                processingTimeMs = elapsedMs(start),
                timestamp = now,
            )
            else -> neutralOutput(name, "No imminent high-impact news (>$CAUTION_MINUTES min out).", start)
        }
    }

    private fun blockOutput(detail: String, startNanos: Long, now: Long) = AgentOutput(
        agentName = name,
        status = AgentStatus.COMPLETE,
        bias = Bias.NEUTRAL,
        confidence = 0.0,
        insights = listOf(
            AgentInsight(
                id = "$name-BLOCK",
                agentName = name,
                type = "NEWS_BLACKOUT",
                direction = null,
                confidence = 100.0,
                timestamp = now,
                detail = detail,
                weight = 2.0,
                tags = listOf("BLOCK"),
            )
        ),
        narrative = "NEWS BLOCK: $detail",
        processingTimeMs = elapsedMs(startNanos),
        timestamp = now,
    )

    private companion object {
        const val BLACKOUT_MINUTES = 15
        const val CAUTION_MINUTES = 60
    }
}

File: reference/typescript-src/agents/impl/news-agent.ts
// ============================================================================
// AGENT 8: NEWS AGENT
// Economic Calendar, Breaking News, Central Banks, Macroeconomics
// ============================================================================

import { TradingAgent, AgentContext, AgentOutput, AgentInsight, AgentName } from '../types';
import { Bias } from '../../core/types';

let seq = 0;

export class NewsAgent implements TradingAgent {
  readonly name: AgentName = 'NEWS';
  readonly description = 'Economic Calendar, Breaking News, Central Banks, Macroeconomics';
  readonly version = '1.0.0';

  analyze(context: AgentContext): AgentOutput {
    const start = performance.now();
    const insights: AgentInsight[] = [];

    // Assess news risk
    const { minutesToHighImpactNews, inNewsBlackout } = context;

    if (inNewsBlackout) {
      insights.push({
        id: `news_${++seq}`, agentName: this.name, type: 'BLACKOUT_ACTIVE',
        direction: null, confidence: 95, timestamp: Date.now(),
        detail: 'High-impact news imminent or just released — extreme volatility expected.',
        weight: 4, tags: ['BLACKOUT', 'EXTREME_RISK'],
      });
    } else if (minutesToHighImpactNews !== undefined && minutesToHighImpactNews < 60) {
      insights.push({
        id: `news_${++seq}`, agentName: this.name, type: 'NEWS_APPROACHING',
        direction: null, confidence: 70, timestamp: Date.now(),
        detail: `High-impact news in ${minutesToHighImpactNews.toFixed(0)} minutes. Plan exits or avoid entries.`,
        weight: 2, tags: ['APPROACHING'],
      });
    } else {
      insights.push({
        id: `news_${++seq}`, agentName: this.name, type: 'NEWS_CLEAR',
        direction: null, confidence: 80, timestamp: Date.now(),
        detail: 'No imminent high-impact news. Calendar is clear for trading.',
        weight: 0.5, tags: ['CLEAR'],
      });
    }

    // News score for probability: 100 = clear, 0 = extreme news risk
    const newsScore = inNewsBlackout ? 10
      : minutesToHighImpactNews !== undefined
        ? Math.min(100, minutesToHighImpactNews * 1.5)
        : 85;

    return {
      agentName: this.name, status: 'COMPLETE', bias: 'NEUTRAL',
      confidence: newsScore, insights,
      narrative: inNewsBlackout
        ? 'NEWS BLACKOUT: Avoid all new entries.'
        : minutesToHighImpactNews && minutesToHighImpactNews < 60
          ? `News in ${minutesToHighImpactNews.toFixed(0)}m. Exercise caution.`
          : 'Calendar clear. Normal trading conditions.',
      processingTimeMs: performance.now() - start, timestamp: Date.now(),
    };
  }

  reset(): void {}
}
